src/main/java/rx/observables/SyncOnSubscribe.java

```diff
@@ -0,0 +1,468 @@
+     * back-pressure. Implementations may not call {@link rx.Subscriber#onNext(Object)
+     * subscriber.onNext(T)} concurrently. You may serialize onNext calls over more than one thread,
+     * but concurrent calls to {@code onNext(t)} will result in a runtime exception. Calling
+     * {@link rx.Subscriber#onCompleted() subscriber.onCompleted()} or
+     * {@link rx.Subscriber#onError(Throwable) subscriber.onError(Throwable)} will result in the
+     * downstream subscriber unsubscribing and executing
+     * {@link SyncOnSubscribe#onUnsubscribe(Object) onUnsubscribe(S)}.
+     * 
+     * @param state
+     *            the value produced by {@link #generateState()} on the first execution or the value
+     *            returned from the previous execution of {@code next(Object, Subscriber)}
+     * @param subscriber
+     *            down stream data subscriber to receive onNext, onCompleted, or onError
+     * @return the state value to be received by the next execution of this function
+     */
+    protected abstract S next(S state, Observer<? super T> subscriber);
+
+    /**
+     * Executed after the downstream subscriber is unsubscribed. This method is useful to override
+     * to clean up the SyncOnSubscribe state. Note that this will not be executed concurrently with
+     * {@link #next(Object, Subscriber) next}.
+     * 
+     * @param state
+     *            the state value
+     */
+    private void onUnsubscribe(S state) {
+        if (onUnsubscribe != null)
+            onUnsubscribe.call(state);
+    }
+    
+    /**
+     * Adds an action to execute after the subscription has been unsubscribed from.
+     * 
+     * @param onUnsubscribe action to execute.
+     * @return this
+     */
+    public SyncOnSubscribe<S, T> onUnsubscribe(Action1<S> onUnsubscribe) {
+        this.onUnsubscribe = onUnsubscribe;
+        return this;
+    }
+
+    /**
+     * Generates a synchronous onSubscribe function that manages back pressure requests delegating
+     * to your {@code next} function/closure to produce data when values are requested by the
+     * downstream subscriber.
+     * 
+     * This overload creates a SyncOnSubscribe without an explicit clean up step.
+     * 
+     * @param generator
+     *            generates the initial state value (see {@link #generateState()})
+     * @param next
+     *            produces data to the downstream subscriber (see {@link #next(Object, Subscriber)
+     *            next(S, Subscriber)})
+     * @return an OnSubscribe that emits data downstream in a protocol compatible with
+     *         back-pressure.
+     */
+    @Experimental
+    public static <S, T> SyncOnSubscribe<S, T> createWithState(Func0<S> generator, final Action2<S, Observer<? super T>> next) {
+        Func2<S, Observer<? super T>, S> nextFunc = new Func2<S, Observer<? super T>, S>() {
+            @Override
+            public S call(S state, Observer<? super T> subscriber) {
```

Comment:
Can we use parameter `observer` rather than `subscriber` now that the type has changed. This is in a few places.